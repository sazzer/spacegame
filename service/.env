RUST_LOG=debug
PORT=8000
DATABASE_URL=postgres://spacegame:spacegame@localhost:45432/spacegame

GOOGLE_CLIENT_ID=748683770071-73hi36dkildfu9ffec5mhv1fdotc2uo3.apps.googleusercontent.com
GOOGLE_CLIENT_SECRET=68pFq1yLhwIND7Z4XehkQCqx
GOOGLE_REDIRECT_URL=http://localhost:8000/authentication/google/complete