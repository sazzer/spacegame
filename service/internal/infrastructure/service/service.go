package service

import "github.com/sirupsen/logrus"

// Service is the actual overall service to run
type Service struct {
}

// NewService builds a new instance of the Service to work with
func NewService() Service {
	logrus.Info("Building Service")

	return Service{}
}

// Start the service running
func (s *Service) Start() {
	logrus.Info("Starting Service")
}
