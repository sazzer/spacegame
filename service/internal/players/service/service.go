package service

// PlayerService is a service structure for interacting with players
type PlayerService struct {
	repository PlayerRepository
}

// NewPlayerService creates a new Player Service
func NewPlayerService(repo PlayerRepository) PlayerService {
	return PlayerService{
		repository: repo,
	}
}
