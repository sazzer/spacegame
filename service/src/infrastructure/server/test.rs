use actix_cors::Cors;
use actix_http::{error::Error, Request};
use actix_service::Service;
use actix_web::http::{HeaderMap, StatusCode};
use actix_web::{middleware::Logger, App};
use bytes::Bytes;
use serde_json::Value;
use std::ops::Deref;

pub struct TestResponse {
  pub status: StatusCode,
  pub headers: HeaderMap,
  pub body: Bytes,
}

impl TestResponse {
  pub fn to_json(&self) -> Result<Value, serde_json::error::Error> {
    serde_json::from_slice(&self.body)
  }
}

pub async fn run_test(server: &super::Server, req: Request) -> Result<TestResponse, Error> {
  let mut app = App::new()
    .wrap(Logger::default())
    .wrap(Cors::new().finish());
  for config in server.configs.iter() {
    app = app.configure(config.deref());
  }

  let mut app = actix_web::test::init_service(app).await;
  let res = app.call(req).await;

  match res {
    Ok(res) => {
      log::info!("Success");
      let status = res.status();
      let headers = res.headers().clone();
      let body = actix_web::test::read_body(res).await;

      Ok(TestResponse {
        status: status,
        headers: headers,
        body: body,
      })
    }
    Err(err) => Err(err),
  }
}
