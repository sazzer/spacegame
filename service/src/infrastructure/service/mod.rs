use super::{
  database::{migrate::migrate_database, Database},
  server::Server,
};
use std::collections::HashMap;
use std::sync::Arc;

/// The settings to use to configure the service
#[derive(Debug, Default)]
pub struct ServiceSettings {
  pub database_url: String,
}

/// The actual service to use
pub struct Service {
  server: Server,
}

impl Service {
  /// Construct a new service
  pub async fn new(settings: ServiceSettings) -> Self {
    log::info!("Building Service");
    let database: Database = Database::new(settings.database_url).await;
    migrate_database(&database).await.unwrap();

    let mut healthchecks: HashMap<String, Arc<dyn crate::infrastructure::health::Component>> =
      HashMap::new();
    healthchecks.insert("database".to_owned(), Arc::new(database.clone()));

    let server = Server::new(vec![
      crate::infrastructure::health::configure::configure_healthchecks(healthchecks),
      crate::authentication::configure::configure_authentication(),
    ]);

    Service { server }
  }

  /// Start the service running
  pub async fn start(&self, port: u16) {
    self.server.start(port).await;
  }

  pub async fn run_test(
    &self,
    req: actix_http::Request,
  ) -> Result<super::server::test::TestResponse, actix_http::error::Error> {
    super::server::test::run_test(&self.server, req).await
  }
}
