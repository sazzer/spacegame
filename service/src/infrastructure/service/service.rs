use std::sync::Arc;

/// The actual Service that does all the work
pub struct Service {
  server: crate::infrastructure::server::Server,
}

impl Service {
  /// Create a new Service
  pub async fn new() -> Self {
    log::info!("Creating Service");

    let db = crate::infrastructure::database::Database::new().await;

    let _player_service = crate::players::configure::new_player_service(db.clone());

    let healthchecker = crate::infrastructure::health::builder::HealthCheckerBuilder::new()
      .with_component("db", Arc::new(db.clone()))
      .build();

    let server = crate::infrastructure::server::Server::new(vec![
      crate::infrastructure::health::configure::configure_healthchecks(healthchecker),
    ]);

    Self { server: server }
  }

  /// Start the service running
  pub async fn start(&self, port: u16) {
    log::info!("Starting Service");
    self.server.start(port).await;
  }
}
