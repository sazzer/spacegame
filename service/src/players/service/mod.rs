mod registration;
mod service;

pub use service::*;
