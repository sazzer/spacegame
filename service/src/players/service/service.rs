use crate::players::database::PlayerRepository;
#[cfg(test)]
use faux;

/// The Player Service to interact with players
#[cfg_attr(test, faux::create)]
#[cfg_attr(not(test), derive(Clone))]
pub struct PlayerService {
  pub(super) repository: PlayerRepository,
}

#[cfg_attr(test, faux::methods)]
impl PlayerService {
  /// Create a new Player Service
  pub fn new(repository: PlayerRepository) -> Self {
    Self { repository }
  }
}
