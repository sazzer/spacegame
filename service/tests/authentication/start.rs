use crate::service::TestService;
use actix_web::test::TestRequest;
use insta::{assert_display_snapshot, assert_json_snapshot};

#[actix_rt::test]
async fn integration_test_start_authentication_invalid() {
  let test_service = TestService::new().await;

  let req = TestRequest::get().uri("/authentication/%20").to_request();

  let res = test_service.request(req).await.unwrap();

  assert_display_snapshot!(res.headers());
  assert_json_snapshot!(res.to_json().unwrap());
}

#[actix_rt::test]
async fn integration_test_start_authentication_unknown() {
  let test_service = TestService::new().await;

  let req = TestRequest::get()
    .uri("/authentication/unknown")
    .to_request();

  let res = test_service.request(req).await.unwrap();

  assert_display_snapshot!(res.headers());
  assert_json_snapshot!(res.to_json().unwrap());
}
