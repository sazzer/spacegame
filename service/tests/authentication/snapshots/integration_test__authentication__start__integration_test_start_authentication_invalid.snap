---
source: tests/authentication/start.rs
expression: res.headers()
---
HTTP/1.1 400 Bad Request
content-type: application/problem+json
